dev/provider_packages/PROVIDER_INDEX_TEMPLATE.rst.jinja2

```diff
@@ -0,0 +1,84 @@
+{#
+ Licensed to the Apache Software Foundation (ASF) under one
+ or more contributor license agreements.  See the NOTICE file
+ distributed with this work for additional information
+ regarding copyright ownership.  The ASF licenses this file
+ to you under the Apache License, Version 2.0 (the
+ "License"); you may not use this file except in compliance
+ with the License.  You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing,
+ software distributed under the License is distributed on an
+ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ KIND, either express or implied.  See the License for the
+ specific language governing permissions and limitations
+ under the License.
+
+ NOTE! THIS FILE IS AUTOMATICALLY GENERATED AND WILL BE
+ OVERWRITTEN WHEN PREPARING PACKAGES.
+
+ IF YOU WANT TO MODIFY IT, YOU SHOULD MODIFY THE TEMPLATE
+ `PROVIDER_INDEX_TEMPLATE.rst.jinja2` IN the `provider_packages` DIRECTORY
+
+#}
+
+Package {{ PACKAGE_PIP_NAME }}
+------------------------------------------------------
+
+Release: {{ RELEASE }}{{ VERSION_SUFFIX }}
+
+Provider package
+----------------
+
+This is a provider package for ``{{PROVIDER_PACKAGE_ID}}`` provider. All classes for this provider package
+are in ``{{FULL_PACKAGE_NAME}}`` python package.
+
+Installation
+------------
+
+.. note::
+
+    On November 2020, new version of PIP (20.3) has been released with a new, 2020 resolver. This resolver
+    does not yet work with Apache Airflow and might lead to errors in installation - depends on your choice
+    of extras. In order to install Airflow you need to either downgrade pip to version 20.2.4
+    ``pip install --upgrade pip==20.2.4`` or, in case you use Pip 20.3, you need to add option
+    ``--use-deprecated legacy-resolver`` to your pip install command.
+
+
+You can install this package on top of an existing airflow 2.* installation via
+``pip install {{PACKAGE_PIP_NAME}}``
+{%- if PIP_REQUIREMENTS %}
+
+PIP requirements
+----------------
+
+{{  PIP_REQUIREMENTS_TABLE_RST | safe}}
+{%- endif %}
+{%- if CROSS_PROVIDERS_DEPENDENCIES %}
+
+Cross provider package dependencies
+-----------------------------------
+
+Those are dependencies that might be needed in order to use all the features of the package.
+You need to install the specified backport providers package in order to use them.
+
+You can install such cross-provider dependencies when installing from PyPI. For example:
+
+.. code-block:: bash
+
+    pip install {{ PACKAGE_PIP_NAME }}[{{ CROSS_PROVIDERS_DEPENDENCIES[0] }}]
+
+
+{{ CROSS_PROVIDERS_DEPENDENCIES_TABLE_RST | safe }}
+
+{%- endif %}
```

Comment:
In my opinion, this page has too much content and some are missing (but the missing ones do not need to be added now).  I like the README files. which is available in the documentation of Google libraries. https://googleapis.dev/python/firestore/latest/index.html
It is standardized too, but contains all the information you need.

1. Starting at the top. So I think it's worth adding a description of the package itself.  This information is available in the `provider.yaml` file, although so far the descriptions are not very. precise.h ttps://github.com/apache/airflow/blob/1602ec97c8d5bc7a7a8b42e850ac6c7a7030e47d/airflow/provider.yaml.schema.json#L13-L16
I hope that we will be able to prepare descriptions in the future gradually, but so far I have made a minimum step and added a link to the product page for each provider.

2. The next step is to link to the documentation, if we want to use the same file in setup.py and in the documentation. This will make it easier to find the documentation when you are in pypi.

We have these links in the side menu, but this is quite a new feature and not everyone uses this menu. Especially on mobile devices, I start reading from the package description, not the side menu.
![Screenshot 2021-01-20 at 03 10 13](https://user-images.githubusercontent.com/12058428/105117662-2924b880-5acd-11eb-987b-445ed4334f9d.png)

3.  The division into a detailed changelog and a normal changelog is strange for me. As for me, the contents of the ADDITIONAL_INFO.rst file should be a normal documentation page and have the title "Migration guide" because this is exactly what this file contains. As this file is fully human written, I don't think other scripts need to process it as well.

4.  I think the detailed changelog and migration guides worth moving to new subpages, because it is not important for the new user.  This then will look similar to the Google documentation
![Screenshot 2021-01-20 at 03 16 53](https://user-images.githubusercontent.com/12058428/105118117-1959a400-5ace-11eb-8d74-1cdc3afe3680.png).
When we move the guides to new pages, we will also be able to easily prepare the migration guide from Airflow 1.10, because we already have a table that facilitates these migrations. 

![Uploading Screenshot 2021-01-20 at 03.21.33.png…]()

5. The links in the table use markdown syntax instead of rst.

Reply:
> In my opinion, this page has too much content and some are missing (but the missing ones do not need to be added now). I like the README files. which is available in the documentation of Google libraries. https://googleapis.dev/python/firestore/latest/index.html
> It is standardized too, but contains all the information you need.
> 
> 1. Starting at the top. So I think it's worth adding a description of the package itself.  This information is available in the `provider.yaml` file, although so far the descriptions are not very. precise.h ttps://github.com/apache/airflow/blob/1602ec97c8d5bc7a7a8b42e850ac6c7a7030e47d/airflow/provider.yaml.schema.json#L13-L16

Sure. Happy to add it. This will be easy to incorporate.

> 2. The next step is to link to the documentation, if we want to use the same file in setup.py and in the documentation. This will make it easier to find the documentation when you are in pypi.

I thiink the PyPI and documentation should be different than documentation. PyPI should be more focused on how to install packages and what are the limitations/requirements, but the "full" documentation" should be reached (including the ChangeLog which I think should not be part of README as it will grow over tim). This is how I approached it and I think this is a better idea than to keep the same documentation index and README.

> We have these links in the side menu, but this is quite a new feature and not everyone uses this menu. Especially on mobile devices, I start reading from the package description, not the side menu.

Yeah. I will modify setup.cfg to include few more links (Documentation especially).
 
> 1. The division into a detailed changelog and a normal changelog is strange for me. As for me, the contents of the ADDITIONAL_INFO.rst file should be a normal documentation page and have the title "Migration guide" because this is exactly what this file contains. As this file is fully human written, I don't think other scripts need to process it as well.

ADDITIONAL_INFO is for backport providers. IMHO this README is something that will evolve evertime. ADDITIONAL_INFO was more about migrating from 1.0 to backports. Effectively CHANGELOG should be written by human as a summary of changes in each version. The assumption is that there will be very rare "migration needed" for those.  Google provider is a bit different than the others, because we migrate python APIs. But in most other packages this will not be a migration guide but human description on what changed in each version. Which I think CHANGELOG is a great name for.

> 2. I think the detailed changelog and migration guides worth moving to new subpages, because it is not important for the new user.  This then will look similar to the Google documentation

Sure. The Detailed guide can be added as a separate page in the docs and linked to. I thought about it as well. Happy to hear why others think but I think we should have this:

* README.md in PyPI describing gerneral "what it is about" + all things important to installing PyPI + link to the Provider "index" documentation

* Provider's index documentation with the same information added + link to detailed commit changelog. It might contain more information (but each Provider is free to add it in the index.rst above the automatically generated part). I do not want to discuss "standard" for this documentation here - we can discuss it separately. In this  change I only want to focus on automatically generated part.

* Detailed changelog - this might be separate page (happy to name it differently if someone has a good proposal).

>    When we move the guides to new pages, we will also be able to easily prepare the migration guide from Airflow 1.10, because we already have a table that facilitates these migrations.
 
We already have it in Backport providers. I think Provider's documentation should be more "forward" looking rather than focuing in the migration from 1.10. 
We might instead point out to backport's PYPI readme and mention that for migration from 1.10 people 

> 1. The links in the table use markdown syntax instead of rst.

I think I fixed that, but I can take a look.

Reply:
> human description on what changed in each version. Which I think CHANGELOG is a great name for.

I am afraid nobody will write it by hand. Google automatically generates CHANGELOG.md files. Everyone has similar references to Github. 
https://github.com/googleapis/python-bigquery/blob/master/CHANGELOG.md
https://github.com/googleapis/python-bigquery/commit/b0e074f7522710886be1da2f117ea22de411b408
To generate the changelog, they use the[googleapis/release-please](https://github.com/googleapis/python-bigquery/commit/b0e074f7522710886be1da2f117ea22de411b408) tool, which analyzes the message and generates a changelog based on it.

> Release Please assumes you are using Conventional Commit messages.
> 
> The most important prefixes you should have in mind are:
> 
> fix: which represents bug fixes, and correlates to a SemVer patch.
> feat: which represents a new feature, and correlates to a SemVer minor.
> feat!:, or fix!:, refactor!:, etc., which represent a breaking change (indicated by the !) and will result in a SemVer major.

In the first version, I think that we don't need the changes to be divided into several categories, although that would be super functional, we can limit ourselves to listing the changes. If we publish these packages regularly, each release will not have a very long list of changes.

Reply:
I like what you have suggested here Jarek.

- Concise README
- Link to commit changelog
- detailed changelog is the Updating guide right?

Reply:
@kaxil For now, I can see that we have two changelogs, one with only a list of breaking changes and the detailed changelog with a list of all commits.
http://gabby-cough.surge.sh/#changelog
We don't have any updating guide yet.

I only suggest changing the name from "Changelog" to "Upgrading guide" and from "Detailed changelog" to "Changelog" because I am concerned that we will be able to keep a changelog written by hand if we have a lot of these contributions. I only suggest changing the name from "Changelog" to "Upgrading guide" and from "Detailed changelog" to "Changelog", because I am concerned that we will be able to maintain the changelog if we have as many contributions as we do now. Over time, it will just copy selected changes that are repeated. in detailed changelog.

Reply:
Oh yes, I agree with that @mik-laj -- My assumption was detailed changelog was updating guide but looks like it is just list of commits.

So yeah, **Changelog** -> **Upgrading guide** and **Detailed changelog** -> **Changelog** .

Reply:
> (happy to name it differently if someone has a good proposal).

Alternatively, we can call it "commits"

Reply:
> > feat!:, or fix!:, refactor!:, etc., which represent a breaking change (indicated by the !) and will result in a SemVer major.

I think automatically generating nice changelog is something we can think about in the future. And if you look at the commit changes now, there are many we would have to exclude because they are just refactors/global updates. For now I thin we should generate changelog manually (and every time someone makes a new feature or breaking change - during review, we should ask the person to add entry to the changelog. I will update the changelogs manually with the first release, but from the 1.0.1 onwards, this is just a matter that all committers will pay attention to it and make sure that Changelog is updated.

As the next step we could introduce the semantic commits and automated changelog generation. But this would require discussion, voting and likely some automation to make sure we follow it. I would like to decouple this from this change.

Reply:
Just one comment here @kaxil @mik-laj about the naming/change process.

I am not totally against naming it UPDATING.md, but I think we are confusing a bit airflow and providers release cycle.  I think we should get the naming right and content of this "manually written" parts.

I believe  most of provider releases will be incremental updates - new hooks/operators added basically or new (backwards-compatible) changes in the operators. The changes that we have now in Google Providers are exception rather than rule. We only need to add the "upgrading" guide because we make backwards-incompatible changes due to 2.0 APIs. In vast majority of cases people should be able to do 'pip install --upgrade `apache-airflow-providers-<NN>` and there should be no need to follow any "updating" guide. This is the main premise of semver, which communicates breaking changes. The only time where we need to add "updating" information is when we go 1.0.0 -> 2.0.0 etc.


Now - this means that vast majority of changes will be simply "new operator "xxx" added". That's why I think CHANGELOG.rst is a good name, and this is also a good name for the Heading in the docs.

While idally indeed we should be able to automaticaly get it from semantic commits (and I am more and more convinced that we should switch to those), this is a decision that will time to discuss and vote and apply and it will take some time for people to adjust to, and we can do it as a follow-up. For now I think this will be very little overhead if we just agree between committers, that every time a new stuff is added to  provider, we ask the PR author to add entry in the CHANELOG.md and provider.yaml:

1) Add new version in provider.yaml patchelevel number (for bugfixes) or minor version (for features)
2) Add a line in CHANGELOG.txt to describe the change added.

I think this makes it nicely distributed among the PR authors, and decreases a lot the work needed by release manager (@kaxil :) - then PR authors and commiters will take care about updating the versions and the role of release manager will be just to review whether everything was included. Remember that release manager will have to sometimes release 10-20 providers at a time and mostly providers that the release manager knows very little about, so the less work is put on the shoulders of that person, the better. 

Let me know WDYT @kaxil @mik-laj and @ashb (I am mentioning @ashb here for the release management part mainly).

Reply:
Also one small comment here.

Looking at https://apache-airflow.slack.com/archives/CCR6P6JRL/p1611225089006700 - this is why I think we should have the detailed commit log (as a separate document linked from the index). Users are sometimes aware of the PRs they are interested in and want to know in which version they were included. And it's not very easy to find it out. We will have a separate set of tags for each provider, so for some people, it would be great to have it in this nicely formatted document rather than having to run some complex git commands to find out.

Reply:
Oh we should definitely have a Changelog which contains the commit log whether it is automated or done manually

Reply:
I will make it separate page then next to index.rst and link to it.

Reply:
I will also add a few hyperlinks here and there, so that we will be able to freely jump between those different pages - 
* README in PyPI, 
* index.rst in docs
* commit changes in docs


Those will be interlinked with version numbers so that we will be able to jump from PYPI 2.0.0 package straight to detailed commit log for 2.0.0 and 2.0.0 Changelog (and all the cross-links between those).